### PR TITLE
ENH adding feature importances support to skpro probabilistic regressors

### PR DESCRIPTION
**Pull request by amanpandey2587**
#### Reference Issues/PRs
Fixes Issue : #376 

#### What does this implement/fix? Explain your changes.
I have added a feature_importances() public method to BaseProbaRegressor in _base.py, along with a new capability:feature_importance tag (default False) in _tags. The method returns a pd.Series of feature importances indexed by feature names seen in fit.
- If the tag is False, calling feature_importances() raises NotImplementedError
- If the tag is True, the method prefers the _feature_importances() private hook over the feature_importances_ attribute,    following existing skpro conventions
- Length of importances is validated against feature_names_in_ and raises ValueError on mismatch

I have also set capability:feature_importance=True in GradientBoostingSurvivalRegressor (_grad_boost_sksurv.py) and implemented the _feature_importances() hook, delegating to the underlying sksurv estimator's feature_importances_ attribute.

#### Does your contribution introduce a new dependency? If yes, which one?
No new dependencies were introduced 

#### What should a reviewer concentrate their feedback on?
- The design of the fallback chain in feature_importances() , here the hook is preferred over attribute so that it is consistent with skpro conventions.
- to check whether the tag name capablility:feature_importance fites the existing tag naming pattern or not
- to check the placement of feature_importances() in _base.py relative to predict()
- test coverage in test_feature_importances,py

#### Did you add any tests for the change?
Yes ,I have added skpro/regression/tests/test_feature_importances.py 
- tag gating :  NotImplementedError when tag is False
- not-fitted guard : RuntimeError before fit
- return type, shape, name, and index alignment with feature_names_in_
- both implementation paths (attribute and _feature_importances() hook)
- error conditions : valueError on length mismatch, AttributeError when tag is True but nothing is implemented
- GradientBoostingSurvivalRegressor integration smoke tests
I am also adding a screenshot of passing of these tests as well 
image

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by amanpandey2587** [comment]
Hi @fkiraly , I have added feature_importances along with tag to NGBoost, GAM, XGBOostLSS , sksurv gradient boosting and CyclicBoosting , also the BaseProbaRefressor.feature_importances now enforces a pd.Series named feature_importance which is indexed be feature_names_in with length checks and docstring explains index ,name and value expectations. and one more thing now the feature_importances supply the standardized Series(hook-first design) so that base now prefers the hook, I have added stub for detection and each model hook returns the standardized pd.seriese which fallback  to feature_importances only if hook is absent .
I have also implemented test for each of them which are passing .
Please check these changes and tell me if anything else is required.
Thanks

File: skpro/regression/gam/_gam.py [hunk 1]
```diff
@@ -209,6 +210,27 @@ def _predict(self, X):
...
+            Index: feature names from ``fit``. Name: ``"feature_importance"``.
+            Values: range (max - min) of partial dependence per term.
+        """
+        gam = self.estimator_
+        names = self.feature_names_in_
+        n = len(names)
+        imp = np.zeros(n, dtype=float)
+        for i in range(n):
+            try:
+                pd_val = gam.partial_dependence(term=i)
```
Comment (fkiraly) [inline skpro/regression/gam/_gam.py]: is this correct if terms are not `"auto"`?

FYI @dswah, your comment would be appreciated.

File: skpro/regression/ensemble/_ngboost.py [hunk 2]
```diff
@@ -92,8 +92,42 @@ class NGBoostRegressor(BaseProbaRegressor, NGBoostAdapter):
...
+        pd.Series
+            Index: feature names from ``fit``. Name: ``"feature_importance"``.
+            Values: from the underlying NGBoost model when base learner supports it.
+        """
+        import numpy as np
+        import pandas as pd
+
+        names = self.feature_names_in_
+        n_features = len(names)
+        if hasattr(self.ngb_, "feature_importances_"):
```
Comment (fkiraly) [inline skpro/regression/ensemble/_ngboost.py]: could you explain when this exists, and point to the exact documentation in ngboost for that?

File: skpro/regression/base/_base.py [hunk 3]
```diff
@@ -204,6 +205,80 @@ def _update(self, X, y, C=None):
...
+        above. Library-specific handling (e.g. mapping from scikit-survival or
+        xgboost) belongs in that override, so the public method always returns
+        the same structure.
+        """
+        self.check_is_fitted()
+
+        if not self.get_tag("capability:feature_importance", False):
+            raise NotImplementedError(
+                f"{type(self).__name__} does not provide feature importances. "
+                "Set the tag 'capability:feature_importance' to True and implement "
```
Comment (fkiraly) [inline skpro/regression/base/_base.py]: I think that is the wrong recommendation for the user. The user should be told that this simply does not exist.
Then call an inner private method.

File: skpro/regression/base/_base.py [hunk 4]
```diff
@@ -204,6 +205,80 @@ def _update(self, X, y, C=None):
...
+                "`_feature_importances` to return a pd.Series in the documented format."
+            )
+
+        hook = getattr(type(self), "_feature_importances", None)
```
Comment (fkiraly) [inline skpro/regression/base/_base.py]: not sure what you are doing and why. I think it should just call a private method, after input checks and conversions. Like every other method.

File: skpro/regression/gam/_gam.py (hunk 1 above)
Comment (amanpandey2587) [reply skpro/regression/gam/_gam.py]: I actually missed that case. The original implementation was using partial_dependence iterating over term indices assuming a 1 to 1 mapping with features.I didn't accounted for cases like interaction terms or explictly passed term structures where teh index i no longer corresponds to feature i.The fix now inspects gam.terms directly, maps each term to its feature indicies via the feature/features attributes and distributes the partial dependence range across all involved features for interaction terms.If term metadata is unavailable it falls back to one term per features. This should handles correctly both cases.

File: skpro/regression/base/_base.py (hunk 3 above)
Comment (amanpandey2587) [reply skpro/regression/base/_base.py]: I have implemented it as per your saying now.

File: skpro/regression/base/_base.py (hunk 4 above)
Comment (amanpandey2587) [reply skpro/regression/base/_base.py]: Same as above, it is now implemented as per your saying

File: skpro/regression/ensemble/_ngboost.py (hunk 2 above)
Comment (amanpandey2587) [reply skpro/regression/ensemble/_ngboost.py]: I had previously added hasattr check as a defensive guard because I had assumed feature_importances_ might not exist on all NGBoost instances ,particularly when using non tree base learners like LinearRegression. My resasoning was based on the sklearn convention where feature_importances_ is only set as an instance attribute when the model supports it. I had also referenced the Base Learners section of the NGBoost usage docs (https://stanfordmlgroup.github.io/ngboost/1-useage.html#base-learners) which documents that any sklearn regressor can be passed as Base but does not describe the internal behaviour of feature_importances_. 
After checking the NGBoost source code at https://github.com/stanfordmlgroup/ngboost/blob/master/ngboost/ngboost.py ,  I found out that feature_importances_ is actually a @property defined directly on the NGBoost base class so it is always present on every fitted instance regardless of the base learner. The property return None when the base leaner is not a tree based regressor. So the hasattr guard was based on a wrong assumption, I should have checked the source code along with docs and sklearn conventions. I had also made the capability:feature_importance to be updated dynamically in _fit after fitting which makes the tag truthful at runtime ( True) only when the fitted base learner actually exposes feature importances
